### PR TITLE
Check whether the state of the ChainDB is in normal form

### PR DESCRIPTION
**Pull request by mrBliss**
Just sharing a potential approach.

I'm not happy with the need  to think about which functions from `NormalForm` to use: `tvar`, `tmvar`, `standard`, etc. We could use type class instances for this (so that we can simply use `custom`, which should then get renamed), but then we'll need to write a *lot* of instances (in most cases `standard` is enough).

I still have to update the mock databases.

There are still a bunch of things not in normal form, causing the test to fail. I have marked these with TODOs. We'll probably need more than a few bangs to fix them.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ImmDB.hs [hunk 1]
```diff
@@ -92,6 +96,16 @@ data ImmDB m blk = ImmDB {
...
     }
 
+instance MonadSTM m => NormalForm m (ImmDB m blk) where
+  custom ImmDB{..} = NF.checkMultiple
+    [ NF.custom immDB
+    , NF.ignore decBlock     -- Parameterised over s
+    , NF.ignore encBlock     -- A function
+    , NF.ignore immEpochInfo -- A record of functions
+    , NF.ignore isEBB        -- A function
+    , NF.ignore err          -- A record of functions
```
Comment (intricate) [inline ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ImmDB.hs]: Just curious: is it not worth asserting that all of these functions/records of functions are in normal form?

Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ImmDB.hs]: Ah, while playing around in GHCi, I had the impression that functions never were in normal form.

The main reason for ignoring these is that they are static anyway, we never update them, so the chances are slim they cause space leaks.

**Comment by edsko** [comment]
Related: #915.

**Comment by intricate** [comment]
This looks like a really cool approach but I wonder whether it would be easier to add support for the inspection of `TVAR` objects in `ghc-heap` (and, as a result, the ability to check whether they're in normal form using `Cardano.Prelude.GHC.Heap.NormalForm.isNormalForm`).

At a first glance, it doesn't seem like it would be _too_ difficult to add the support for inspecting `TVAR`s in `ghc-heap` itself. The [`StgTVar` struct in GHC contains a pointer to the actual object that it's wrapping](https://gitlab.haskell.org/ghc/ghc/blob/672cbab268342d7a8829698e838e369e0a0b3a19/includes/rts/storage/Closures.h#L324), which should be all we need in order to check whether it's in normal form. So it _seems_ simple to add support for `TVAR` objects but, if it's really as simple as I'm thinking, I have to wonder why it hasn't be done already (as I've realized, things aren't always as simple as they seem :stuck_out_tongue:).

However, if we go this route, it would take some time to get this merged into `ghc-heap` and released, so maybe it'd be a good idea to follow this approach and swap it out once there's `ghc-heap` support for STM objects. Or perhaps we could temporarily include the solution in `cardano-prelude` (if possible)?

**Comment by edsko** [comment]
I think they are complimentary. `StrictTVar` (and co) mean that we get type-level guarantees that `TVar`s are in weak head normal form. Type level guarantees are of course always better than testing :) _But_ they only guarantee _weak head_ normal form, not normal form. Extending the `isNormalForm` stuff to cover `TVar`s would still be _very_ useful to check the latter. I think we should have both.

**Comment by edsko** [comment]
I agree that it would be very worthwhile to extend the `isNormalForm` to cover `TVars`. Combined with some type level guarantees that `TVar`s will be in weak head normal form (#918), it will get us quite far I think.

**Comment by edsko** [comment]
Infrastructure obsoleted by https://github.com/input-output-hk/cardano-prelude/pull/67 , but there might still be some special cases in this PR that we should keep.

**Comment by edsko** [comment]
Replacement PR: #1039 .

**Pull request closed without merging**